### PR TITLE
Added PXE UEFI support to 810_create_pxelinux_cfg.sh (issue 2359)

### PR DESCRIPTION
**Pull request by jsmeix**
* Type: **Enhancement**

* Impact: **High**

* Reference to related issue (URL):
https://github.com/rear/rear/issues/2359

* How was this pull request tested?
Not at all tested by me because I do not use PXE.

* Brief description of the changes in this pull request:

Manually added the additional lines from the diff output in
https://github.com/rear/rear/issues/2359#issuecomment-615197886

### CODE REVIEW

File: usr/share/rear/output/PXE/default/810_create_pxelinux_cfg.sh [hunk 1]
```diff
@@ -32,19 +32,50 @@ if [[ ! -z "$PXE_CONFIG_URL" ]] ; then
...
+        cat >"$PXE_LOCAL_PATH/$PXE_CONFIG_FILE" <<EOF
+set menu_color_normal=white/black
+set menu_color_highlight=black/light-gray
+set timeout=30
+insmod png
+insmod efi_gop
+insmod efi_uga
+insmod video_bochs
+insmod video_cirrus
+insmod all_video
```
Comment (jsmeix) [inline usr/share/rear/output/PXE/default/810_create_pxelinux_cfg.sh:48]: I think enfore loading the GRUB2 modules
efi_gop, efi_uga, video_bochs, and video_cirrus
is obsolete when the generic "insmod all_video" is there
and enfore loading the other GRUB2 modules lets things fail
on systems where those other modules do not exist,
cf. https://github.com/rear/rear/issues/2388

File: usr/share/rear/output/PXE/default/810_create_pxelinux_cfg.sh [hunk 2]
```diff
@@ -32,19 +32,50 @@ if [[ ! -z "$PXE_CONFIG_URL" ]] ; then
...
-    display $OUTPUT_PREFIX_PXE/$PXE_MESSAGE
-    say ----------------------------------------------------------
-    say rear = disaster recover this system with Relax-and-Recover
-    label rear
-	kernel $OUTPUT_PREFIX_PXE/$PXE_KERNEL
-	append initrd=$OUTPUT_PREFIX_PXE/$PXE_INITRD root=/dev/ram0 vga=normal rw $KERNEL_CMDLINE $PXE_RECOVER_MODE
+    # UEFI case, cf. https://github.com/rear/rear/issues/2359
+    EFIMT="$( mount -v | grep /sys/firmware/efi/efivars | wc -l )"
+    EFIMGR="$( efibootmgr -v | grep "BootOrder" | wc -l )"
+    if [[ x$EFIMT = x1 && x$EFIMGR = x1 ]] ; then
```
Comment (jsmeix) [inline usr/share/rear/output/PXE/default/810_create_pxelinux_cfg.sh:38]: That `[[ x$EFIMT = x1 && x$EFIMGR = x1 ]]` should be
replaced by more usual coding style.

File: usr/share/rear/output/PXE/default/810_create_pxelinux_cfg.sh [hunk 3]
```diff
@@ -32,19 +32,50 @@ if [[ ! -z "$PXE_CONFIG_URL" ]] ; then
...
-    $(test -s $(get_template "PXE_pxelinux.cfg") && cat $(get_template "PXE_pxelinux.cfg"))
-    display $OUTPUT_PREFIX_PXE/$PXE_MESSAGE
-    say ----------------------------------------------------------
-    say rear = disaster recover this system with Relax-and-Recover
-    label rear
-	kernel $OUTPUT_PREFIX_PXE/$PXE_KERNEL
-	append initrd=$OUTPUT_PREFIX_PXE/$PXE_INITRD root=/dev/ram0 vga=normal rw $KERNEL_CMDLINE $PXE_RECOVER_MODE
+    # UEFI case, cf. https://github.com/rear/rear/issues/2359
+    EFIMT="$( mount -v | grep /sys/firmware/efi/efivars | wc -l )"
+    EFIMGR="$( efibootmgr -v | grep "BootOrder" | wc -l )"
```
Comment (jsmeix) [inline usr/share/rear/output/PXE/default/810_create_pxelinux_cfg.sh:37]: I wonder if there exists a "BootOrder" entry in any case when UEFI is used?
I think the UEFI firmware can boot a default/fallback EFI binary
basically without any UEFI configuration entries.

File: usr/share/rear/output/PXE/default/810_create_pxelinux_cfg.sh [hunk 4]
```diff
@@ -32,19 +32,50 @@ if [[ ! -z "$PXE_CONFIG_URL" ]] ; then
...
-    cat >"$PXE_LOCAL_PATH/$PXE_CONFIG_FILE" <<EOF
-    $(test -s $(get_template "PXE_pxelinux.cfg") && cat $(get_template "PXE_pxelinux.cfg"))
-    display $OUTPUT_PREFIX_PXE/$PXE_MESSAGE
-    say ----------------------------------------------------------
-    say rear = disaster recover this system with Relax-and-Recover
-    label rear
-	kernel $OUTPUT_PREFIX_PXE/$PXE_KERNEL
-	append initrd=$OUTPUT_PREFIX_PXE/$PXE_INITRD root=/dev/ram0 vga=normal rw $KERNEL_CMDLINE $PXE_RECOVER_MODE
+    # UEFI case, cf. https://github.com/rear/rear/issues/2359
+    EFIMT="$( mount -v | grep /sys/firmware/efi/efivars | wc -l )"
```
Comment (jsmeix) [inline usr/share/rear/output/PXE/default/810_create_pxelinux_cfg.sh:36]: I wonder if /sys/firmware/efi/efivars is mounted in any case when UEFI is used.

**Comment by gozora** [comment]
I will certainly take a look, but I need to setup PXE testing environment first, which will take some time ...
@didacog aren't you guys heavily using PXE in DRLM ? Did you noticed some problems with UEFI based systems?

V.

**Comment by didacog** [comment]
Hi @gozora,
Yes we use PXE boot all the time in DRLM with both BIOS and UEFI systems with no issues.

**Comment by jsmeix** [comment]
@didacog 
could you describe how your ReaR PXE setup for UEFI
is different compared to
the "ReaR PXE setup for RHEL/OEL 7 to boot with UEFI"
in https://github.com/rear/rear/issues/2359
together with the also needed additional code in this pull request here?

In particular what is your etc/rear/local.conf
where PXE boot with UEFI works in your case?

**Comment by didacog** [comment]
Hi,
We just configure a MACAddr file to setup the boot and adjust de vga to gfxpayload=vga of the kernel opts. See: https://github.com/brainupdaters/drlm/blob/develop/usr/share/drlm/backup/run/default/129_post_backup_tasks.sh

You can also take a look at out GRUB2 config on the server side: https://github.com/brainupdaters/drlm/blob/develop/var/lib/drlm/store/boot/grub/grub.cfg.

Its a mkstandalone GRUB image BTW in DRLM.

You can take a look in how we created it at: https://docs.drlm.org/en/2.3.1/building_grub2.html

Regards,
Didac

**Comment by jsmeix** [comment]
As far as I understand https://github.com/rear/rear/pull/2449#issuecomment-655976230
it means that DRLM uses its own scripts to set up PXE boot with UEFI
so PXE boot with UEFI can "just work" in DRLM while it can "just fail" in ReaR.

**Pull request closed without merging**